### PR TITLE
Add Avaje Inject Module

### PR DESCRIPTION
**Pull request by SentryMan**
First pass at an avaje inject module

helps alleviate #3415

### CODE REVIEW

File: modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/JoobyPropertyPlugin.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  @Override
+  public boolean contains(String property) {
+    return config.hasPath(property);
+  }
+
+  @Override
+  public boolean equalTo(String property, String value) {
+    return config.getString(property).equals(value);
```
Comment (edgarespinawt) [inline modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/JoobyPropertyPlugin.java:29]: should we check for existence first?

**Review by edgarespinawt — APPROVED** [review]
Good addition, thank you. Just one comment

**Review by ludmiloff — COMMENTED** [review]
I grabbed the source and tried locally. It results in multiple Hikari re-initialisation and database connections and application crash. This is due to ```beanScope.provideDefault(key.getName(), key.getType(), e::getValue);``` which adds the bean as secondary. 
I suggest changing it to ```beanScope.bean(key.getType(), e.getValue());```.

With this change my testing code works, still needs ```@InjectModule(requires)``` everywhere.

**Review by ludmiloff — CHANGES_REQUESTED** [review]
Please, change service registration as follows:
```java
if (key.getName() == null) {
  beanScope.provideDefault(key.getType(), e::getValue);
} else {
  beanScope.bean(key.getName(), key.getType(), e.getValue());
}
```

This:
1) prevents adding multiple secondary beans for same type, Avaje complains about that
2) allows named injection, e.g. ```@Named("db1") Jdbi``` 

ps: adding package-info.java with specific ```InjectModule(name = 'jooby-modules', requires = {...})``` is required for pass avaje-generator compillation

**Comment by SentryMan** [comment]
@ludmiloff how's this?

**Review by ludmiloff — CHANGES_REQUESTED** [review]
As @jknack mentioned before, there must be an existence check in JoobyPropertyPlugin, line 30
@SentryMan, may I suggest the change like this:
```java
return config.hasPath(property) && Objects.equals(config.getString(property), value);```

File: modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * <p>Jooby integrates the {@link io.jooby.ServiceRegistry} into the Avaje DI framework.
+ *
+ * <p>Usage:
+ *
+ * <pre>{@code
+ * {
+ *
+ *
+ *   install(new AvajeInjectModule());
```
Comment (jknack) [inline modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java:29]: should be: `install(AvajeInjectModule.of())` or we need to make constructor public

File: modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  AvajeInjectModule(BeanScopeBuilder beanScope) {
+    this.beanScope = beanScope;
+  }
+
+  @Override
+  public void install(Jooby application) throws Exception {
+
+    application
```
Comment (jknack) [inline modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java:59]: This probably better to run on `starting` or `started`. So we don't depend of sequential order of services being added before this module.

Comment (SentryMan) [reply modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java:64]: would lateInit=true resolve?

**Review by ludmiloff — COMMENTED** [review]
@SentryMan, your latest commit about named property injection works in the following scenario:

```java
@Inject @Named("myProp") String myProp
```

and the actual value is available during/after PostConstruct phase.

Constructor injection of named config item (Jooby 1 style, still good :) ) does not work, however.

Existence check in JoobyPropertyPlugin is fine.

**Review by ludmiloff — COMMENTED** [review]
A bit of explanation about my latest comment:
constructor injection if named config item does not work because it needs to be declared as external dependency, which might be very cumbersome.

**Comment by SentryMan** [comment]
you mean with the @InjectModule(requires) thing?

File: modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java (hunk 3 above)
Comment (ludmiloff) [reply modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java:64]: Tested, with lateinit=true the order does not matter anymore.

**Comment by ludmiloff** [comment]
> you mean with the `@InjectModule(requires={String.class})` thing? For that, you only need to say `String.class` once and it'll take care of things. I'll need to make a note of that in the docs.

Yes, this way it works, thanks.

File: modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java (hunk 3 above)
Comment (jknack) [reply modules/jooby-avaje-inject/src/main/java/io/jooby/avaje/inject/AvajeInjectModule.java:64]: right. lateinit will fix it

**Review by ludmiloff — COMMENTED** [review]
I think this module is already done, at least for a sort of EAP. 
My notes about:
1. Needs documentation
2. Injecting property (incl. com.typesafe.config.Config items by name) works
3. Constructor injection works
4. All JoobyModule(s) currently in use are injected as Avaje Beans, named injection is also supported
5. All Avaje Beans are accessible via Jooby require
6. Installing AvajeInjectModule does not need a specific order in code along other Jooby modules
7. Packaging my test project into a single jar produces a working deployment without any problem.

Some limitation and caveats:
1. Using this module needs a package-info.java with proper InjectModule(require) clause. This is very Avaje specific.
2. Config item injection is currently limited to String types or string representation (toString). Might be improved with Integer and Boolean classes, though.
3. Using this module needs a special maven/gradle plugin and setup for Avaje code generation
4. Using this module adds a small overhead in startup time, on my testing machine 8 cores M1, java 17, small size project with 3 other jooby modules and 200 lines of config items, cold start in dev mode - around 20ms for wiring beans at runtime (as reported by Avaje), hope this is acceptable and much better than Guice

Finally, this is ready for approval and merging, IMO.
@jknack, @SentryMan, thank you for your support.

**Comment by SentryMan** [comment]
> 3. Using this module needs a special maven/gradle plugin and setup for Avaje code generation

This should only be required if you are running on the module-path and also happen to have custom avaje autoconfig libraries. But yeah it is what it is

**Pull request merged**